### ISSUE TITLE
Enforce ExternalID uniqueness

### ISSUE THREAD

**Issue by nilebox** [kubernetes-incubator/service-catalog#2035]
Service Catalog allows creating two `ServiceInstance` objects with explicit duplicate `ExternalID`. That doesn't seem right? Is there a way to enforce `ExternalID` uniqueness in Kubernetes?

We can at least have an index in the informer and have a check for duplicates in the beginning of reconciliation loop and fail fast if we detect an instance with `ExternalID`.

**Comment by nilebox** [comment]
If we don't address this issue, I see 2 potential problems:
1. We could have 2 `ServiceInstance` objects and one of them will be out of sync (i.e. race, last update wins)
2. Someone could deliberately create a `ServiceInstance` pointing to existing instance owned by the other user to mess with it. See #1928

**Comment by duglin** [comment]
Do we allow people to set that? If so, why?  But I agree,we should not allow dups.

**Comment by nilebox** [comment]
@duglin yes, `ExternalID` can be provided by the user. If it's not provided, Service Catalog will generate a new UUID.

Allowing user to provide `ExternalID` is useful for migrating an existing instance. I.e. it's a no-op from the OSB side, but Service Catalog will know about the existing instance that was provisioned some other way (e.g. via Cloud Foundry or the other instance of Service Catalog) previously.

**Comment by duglin** [comment]
seems really risky to me since that now switches these fields from "some random uuid that really only matters to this svc-cat to get its job done", to "a uuid that could possibly have meaning outside of this svc-cat".  But ok...

**Comment by nilebox** [comment]
@duglin it could make sense to have a flag for enabling/disabling support for user-provided `ExternalID`? That actually could be useful for one-time migration onto Service Catalog with the flag enabled and then disabling it. That would work for us, interested to learn about other people's scenarios.

**Comment by nilebox** [comment]
Marking the issue with `1.0.0.` milestone, I think it's important enough to address before GA.

**Comment by wryun** [comment]
I think a flag makes sense. In normal use cases, Service Catalog should be the 'owner' (and sole authority) on the resource, and it's dangerous for the user to interact with the UUID. It would still be nice to prevent duplicates, though...

On the other hand, Service Catalog forbidding duplicates doesn't prevent all problems: if you have separate Service Catalogs referring to the same broker, they can clash. One option would be to encourage brokers to persist the context field (i.e. in this case, clusterID would distinguish) to prevent this.

**Comment by wryun** [comment]
Thinking a little more about this, maybe the default behaviour should be to use the kubernetes UID as the ExternalID. Then, if you flag on 'external control', ExternalID is a required field - and if they need it to be autogenerated in only some cases, they should have a MutatingAdmissionWebhook  (i.e. make it so there is a single spot that owns the UUID creation in either case - either Kube or the user).

**Comment by duglin** [comment]
In general how does Kube deal with migrations? Do people really use kubectl to do that and have to go thru all of the normal validity & rbac checks? Who would own all of the resources at that point? Or is migration more of an etcd migration/reload thing?  I would think a migration, backup/restore, type of thing would be more at the DB level.  Additionally, does Kube support migrations of just certain resources?

**Comment by wryun** [comment]
In our case, we're migrating existing resources to Service Catalog on user request (i.e. it's a other -> SC, not SC -> SC). The actual migration time is driven by the user (it's not a bulk move), so it's easier to say 'just specify your existing UUID' than develop a separate process. Also, going through the front-door guarantees that we get valid records, as opposed to developing our own method of constructing them in etcd.

In terms of how Kubernetes deals with migration, don't know; my suspicion is that the Service Catalog use case is fairly different from standard Kubernetes objects (where the main thing you'd be migrating between clusters would be the persistent storage, and everything else would just be regenerated by your deployment system).

**Comment by mrbobbytables** [comment]
This project is being archived, closing open issues and PRs. 
Please see this PR for more information: https://github.com/kubernetes/community/pull/6632
